File: chp03_AbstractFactory/Kotlin/ThemedApplication_basic.kt
// Abstract Products
interface Button {
    fun render()
}

interface TextBox {
    fun render()
}

// Concrete Products for Light Theme
class LightButton : Button {
    override fun render() {
        println("Rendering a Light Button")
    }
}

class LightTextBox : TextBox {
    override fun render() {
        println("Rendering a Light TextBox")
    }
}

// Concrete Products for Dark Theme
class DarkButton : Button {
    override fun render() {
        println("Rendering a Dark Button")
    }
}

class DarkTextBox : TextBox {
    override fun render() {
        println("Rendering a Dark TextBox")
    }
}

// Abstract Factory
interface ThemeFactory {
    fun createButton(): Button
    fun createTextBox(): TextBox
}

// Concrete Factories for Light Theme
class LightThemeFactory : ThemeFactory {
    override fun createButton(): Button {
        return LightButton()
    }

    override fun createTextBox(): TextBox {
        return LightTextBox()
    }
}

// Concrete Factories for Dark Theme
class DarkThemeFactory : ThemeFactory {
    override fun createButton(): Button {
        return DarkButton()
    }

    override fun createTextBox(): TextBox {
        return DarkTextBox()
    }
}

// Client Code
class ThemedApplication(private val factory: ThemeFactory) {
    fun renderUI() {
        val button = factory.createButton()
        val textBox = factory.createTextBox()
        button.render()
        textBox.render()
    }
}

// Usage
fun main() {
    println("Using Light Theme:")
    val lightThemeApp = ThemedApplication(LightThemeFactory())
    lightThemeApp.renderUI()

    println("\nUsing Dark Theme:")
    val darkThemeApp = ThemedApplication(DarkThemeFactory())
    darkThemeApp.renderUI()
}

File: chp03_AbstractFactory/Swift/ThemedApplication_basic.swift
import Foundation

// Abstract Products
protocol Button {
    func render()
}

protocol TextBox {
    func render()
}

// Concrete Products for Light Theme
class LightButton: Button {
    func render() {
        print("Rendering a Light Button")
    }
}

class LightTextBox: TextBox {
    func render() {
        print("Rendering a Light TextBox")
    }
}

// Concrete Products for Dark Theme
class DarkButton: Button {
    func render() {
        print("Rendering a Dark Button")
    }
}

class DarkTextBox: TextBox {
    func render() {
        print("Rendering a Dark TextBox")
    }
}

// Abstract Factory
protocol ThemeFactory {
    func createButton() -> Button
    func createTextBox() -> TextBox
}

// Concrete Factory for Light Theme
class LightThemeFactory: ThemeFactory {
    func createButton() -> Button {
        return LightButton()
    }
    
    func createTextBox() -> TextBox {
        return LightTextBox()
    }
}

// Concrete Factory for Dark Theme
class DarkThemeFactory: ThemeFactory {
    func createButton() -> Button {
        return DarkButton()
    }
    
    func createTextBox() -> TextBox {
        return DarkTextBox()
    }
}

// Client Code
class ThemedApplication {
    private let factory: ThemeFactory
    
    init(factory: ThemeFactory) {
        self.factory = factory
    }
    
    func renderUI() {
        let button = factory.createButton()
        let textBox = factory.createTextBox()
        
        button.render()
        textBox.render()
    }
}

// Usage
print("Using Light Theme:")
let lightApp = ThemedApplication(factory: LightThemeFactory())
lightApp.renderUI()

print("\nUsing Dark Theme:")
let darkApp = ThemedApplication(factory: DarkThemeFactory())
darkApp.renderUI()
